Getting rid of _make_points_and_values_ascending and its unnecessary sorting, refactoring

ideasrule opened the pull request:
Got rid of _make_points_and_values_ascending, which unnecessarily sorts the input.

Pulled _check_points and _check_dimensionality out of RegularGridInterpolator so that they can also be used by interpn.  This reduced code duplication, in addition to helping get rid of the aforementioned _make_points_and_values_ascending.

ev-br commented:
Looks good to me!
pythran error on musl CI does not seem related.

@kmuehlbauer heads up.

ev-br commented:
Oh, and for future PRs (no action needed now) : sending PRs from the main branch gets very messy very fast, so you'll save yourself a fair bit of git voodoo practice if you make a new branch for a PR.

tylerjereddy reviewed (commented):
I think there's a `class`-associated method you can delete now as well. The full test suite still passes for me on this branch locally when I apply the diff below. If that could be pushed in soon we can merge this before the next release branch, which is quite soon..

```diff
diff --git a/scipy/interpolate/_rgi.py b/scipy/interpolate/_rgi.py
index ca88c3668..cfc8b16ed 100644
--- a/scipy/interpolate/_rgi.py
+++ b/scipy/interpolate/_rgi.py
@@ -243,17 +243,6 @@ class RegularGridInterpolator:
         if self._descending_dimensions:
             self.values = np.flip(values, axis=self._descending_dimensions)
 
-    def _check_dimensionality(self, points, values):
-        if len(points) > values.ndim:
-            raise ValueError("There are %d point arrays, but values has %d "
-                             "dimensions" % (len(points), values.ndim))
-        for i, p in enumerate(points):
-            if not np.asarray(p).ndim == 1:
-                raise ValueError("The points in dimension %d must be "
-                                 "1-dimensional" % i)
-            if not values.shape[i] == len(p):
-                raise ValueError("There are %d points and %d values in "
-                                 "dimension %d" % (len(p), values.shape[i], i))
 
     def _check_values(self, values):
         if not hasattr(values, 'ndim'):
```

ideasrule commented:
> I think there's a `class`-associated method you can delete now as well. The full test suite still passes for me on this branch locally when I apply the diff below. If that could be pushed in soon we can merge this before the next release branch, which is quite soon..

Oops, I totally forgot to delete that!  But now that I've deleted it, I notice one problem.  In the benchmarking code (benchmarks/benchmarks/interpolate.py), there's a subclass of RegularGridInterpolator where the __call__ method is:

```
def __call__(self, values, method=None):
        values = self._check_values(values)
        # check fillvalue
        self._check_fill_value(values, self.fill_value)
        # check dimensionality
        self._check_dimensionality(self.grid, values)
        # flip, if needed
        self.values = np.flip(values, axis=self._descending_dimensions)
        return super().__call__(self.xi, method=method)
```

The self._check_dimensionality line now fails.  I have no idea why all of these checks are there in the first place, but I don't want to mess with the benchmarking code right now, so I added wrapper class methods for _check_dimensionality and _check_points.  This also has the side benefit of making the code more elegant and intuitive, because previously, _check_values was a class method but the other two checking functions were not.

ev-br commented:
The subclass in the benchmark suite came in in https://github.com/scipy/scipy/pull/17230 (there is a long discussion in it and its previous iterations). 

In particular ( https://github.com/scipy/scipy/pull/17230#issuecomment-1288777273),

> Maybe it's best to closely follow scipy development and keep in sync that way.

which is why I pinged @kmuehlbauer upthread. Let's give it a couple of days for him to hopefully respond before 1.10 is branched.

kmuehlbauer commented:
@ev-br Thanks for pinging.

@ideasrule The work in #17230 was merely to make RGI subclassable without too much effort for our use case. It looks I might have done better. The wrappers are good to keep it subclassable.

The subclass is taking input and output grid information on initialization and values on call. So all these checks are needed in the subclass call to verify proper conditions of values. The grids are checked on initialization.

Hope that gives some idea of the whereabouts.

ev-br commented:
Okay, this looks good, thanks for @ideasrule , thanks @kmuehlbauer @tylerjereddy for reviews and suggestions. Merging.
